lingk_scraper.py

```diff
@@ -0,0 +1,155 @@
+                    iterations: int=10) -> dict:
+    '''
+    Returns the JSON representation of the course data.
+    iterations: the number of iterations to try connecting
+                (it sometimes doesn't authenticate).
+    Credit to Jordan R.A., Olivia Watkins, Kate Woolverton, and Gavin Yancey
+    for figuring out the nuances of this awful, broken API.
+    '''
+    for _ in range(iterations):
+        date_str = time_html_format()
+        auth_header = gen_auth_header(key, secret, date_str)
+        # they want the header nested inside of this for some reason...
+        headers = {'Date': date_str, 'Authorization': auth_header}
+        params = {'limit': limit}
+        r = requests.get(f'https://{BASE + EXTENSION}',
+                         headers=headers, params=params)
+        if r.status_code == requests.codes.ok:
+            return r.json()['data']
+    else:
+        log(f'lingk scraping failed after {iterations} iterations')
+
+# split lingk entityID to department, course_number, num_suffix, school
+# 'SPAN199DRPO' => SPAN, 199, DR, PO
+# 'ENGR190AKHM' => ENGR, 190, AK, HM
+# 'FGSS189  SC' => FGSS, 189, , SC
+COURSE_ID_REGEX = r"([A-Z]+) *([0-9]+) *([A-Z]*[0-9]?) *([A-Z]{2})"
+
+def make_description_dict(lingk_data: list) -> dict :
+    '''
+    Make a dictionary that points the course key in the format
+    'department/course_number/suffix/school' to the corresponding description
+    string scrpaed from lingkData. The lingkData is an array of courses scraped 
+    from lingk api.
+
+    Note: if the lingkData is not properly filtered, the dict might contain everything
+    since 2016.
+    '''
+    lingk_description_dict = {}
+    for course in lingk_data :
+        try:
+            # Many courses in the lingk course list do not have descriptions.
+            # Many of them are malformed.
+            if 'description' in course :
+                # There is a course in the lingk api that looks like this:
+                # "externalId": "\n\"FGSS189  SC" this line is just to account
+                # for that.
+                rawstr = course['externalId'].strip().strip('\"')
+
+                match = re.match(COURSE_ID_REGEX, rawstr)
+                if not match :
+                    raise ScrapeError("\"%s\" : lingk malformed course, string not match" % rawstr)
+                department, course_number, num_suffix, school = match.groups()
+                hyperscheduleKey = '/'.join([department, course_number, num_suffix, school])
+                # course description from lingk sometimes have multiple newlines
+                description = course['description'].strip()
+                if hyperscheduleKey in lingk_description_dict :
+                    raise ScrapeError("%s : lingk duplicate course key" % hyperscheduleKey)
+                lingk_description_dict[hyperscheduleKey] = description
+        except Exception as e:
+            log(type(e).__name__ + " " + "Exception:" + " " + str(e))
```

Comment:
This is not adequate error handling. We should only do a soft failure in the cases where the input data is actually malformed.